### PR TITLE
travis: turn on -Werror for make

### LINKED ISSUE #776: Eliminate compiler warnings and turn on -Werr
I used the master branch.

```
../modules/afprog/afprog.c: In function 'afprogram_reload_store_item_free':
../modules/afprog/afprog.c:69:3: warning: passing argument 1 of 'log_pipe_unref' from incompatible pointer type [enabled by default]
   log_pipe_unref(reload_info->writer);
   ^
In file included from ../lib/driver.h:29:0,
                 from ../modules/afprog/afprog.h:27,
                 from ../modules/afprog/afprog.c:24:
../lib/logpipe.h:237:6: note: expected 'struct LogPipe *' but argument is of type 'struct LogWriter *'
 void log_pipe_unref(LogPipe *self);
      ^
../modules/afprog/afprog.c: In function 'afprogram_dd_kill_child':
../modules/afprog/afprog.c:329:13: warning: unused variable 'pgid' [-Wunused-variable]
       pid_t pgid;
             ^
../modules/afprog/afprog.c: In function 'afprogram_dd_store_reload_store_item':
../modules/afprog/afprog.c:452:3: warning: passing argument 4 of 'cfg_persist_config_add' from incompatible pointer type [enabled by default]
   cfg_persist_config_add(cfg, afprogram_dd_format_persist_name(self), reload_info, afprogram_reload_store_item_destroy_notify, FALSE);
   ^
In file included from ../lib/logpipe.h:30:0,
                 from ../lib/driver.h:29,
                 from ../modules/afprog/afprog.h:27,
                 from ../modules/afprog/afprog.c:24:
../lib/cfg.h:158:6: note: expected 'GDestroyNotify' but argument is of type 'void (*)(struct AFProgramReloadStoreItem *)'
 void cfg_persist_config_add(GlobalConfig *cfg, gchar *name, gpointer value, GDestroyNotify destroy, gboolean force);

```

To do:
- [x] Eliminate compiler warnings - #779
- [ ] Turn on -Werr #871

### LINKED ISSUE #804: add --disable-werr option
#776 is about to enable -Werr for syslog-ng. Making compilation warning free is done on Ubuntu. I do packaging on other distros with many different releases. Out of experience these give warnings at very different places. Please add a --disable-werr option, so I can compile packages anywhere. I'm happy to provide you with warning messages if I need to enable this option any time.

### PR DESCRIPTION
**Pull request by ihrwein**
Fixes #804 
Fixes #776

Signed-off-by: Tibor Benke <tibor.benke@balabit.com>

### CODE REVIEW

**Comment by ihrwein** [comment]
@bazsi @juhaszviktor could you review this PR?

**Comment by juhaszviktor** [comment]
:+1:

**Comment by ihrwein** [comment]
Update: I added a `-1` substraction to the `strncat` call.

**Comment by MrAnno** [comment]
:+1:

**Comment by ihrwein** [comment]
@lbudai Could you merge this?

**Comment by bkil-syslogng** [comment]
I would +1, if you #undef'ed the macros in conflict. Leaking out of such definitions from headers does usually indicate a bug in my opinion. You could consider #ifdef/#undef if you think this is platform dependent. However, if this is a bug in one of the libraries, we should also report towards them.

**Comment by bkil-syslogng** [comment]
So instead of all the hacky changes in the last commit, I would simply alter python-module.h something like this:
```
#include "syslog-ng.h"

#ifdef _POSIX_C_SOURCE
#include <Python.h>
#else
#include <Python.h>
#undef _POSIX_C_SOURCE
#endif
```

I guess you would need some combinatorics as well if multiple macros are involved, or somehow save the previous value. Those should be given through the compiler and/or command line, and usually not from intermediate code I think.

**Comment by bkil-syslogng** [comment]
An alternative would be to define the said macros at the very beginning of syslog-ng.h if we or our dependencies really take advantage of them. We should then introduce a policy that every file must include syslog-ng.h before any other header file (I guess this is mostly the case at the moment anyhow):

http://pubs.opengroup.org/onlinepubs/007904975/functions/xsh_chap02_02.html

**Comment by bkil-syslogng** [comment]
TL;DR:
```
if test "x$enable_python" = "xyes" -a "x$python_found" = "xyes"; then
  CFLAGS="${CFLAGS} -D_XOPEN_SOURCE=500 -D_POSIX_C_SOURCE=200112L"
fi
```
I have a still little better workaround. The below patch is only a quickie not my final proposal, but at least it could give you some new ideas:
https://github.com/bkil-syslogng/syslog-ng/commit/e8ea0e25002aca26d03b7f225b154237c7859abb

**Comment by bazsi** [comment]
I think we could make an exception with the Python header here, and include
it before any other header. Clearly it was not entirely designed as a
library header. IIRC this problem is also fixed with newer Python versions.
I hope I am not missing anything.

**Comment by bkil-syslogng** [comment]
@bazsi I think we agree that this is a Python bug. I'm not sure which
versions this plagues. What makes me concerned is that certain versions of
this header can leak this macro outside, while others do not. In this case,
it taints our libraries, modules and all of our users to who this
propagates to. This could introduce unwanted behavior changes in
implementation.

What about the principle of least surprise and documenting intention by
code? If we silently place the `#include <Python.h>` to the top and also
alter all includes to which this reaches, we're not documenting a bug to
watch out for in the future.

How about this compromise? We place something like `#ifdef WITH_PYTON /
#include <Python.h>` at the very beginning of syslog-ng.h and introduce a
policy that `#include "syslog-ng.h"` must come first everywhere and must be
present in all files which `#include` anything.

**Comment by bkil-syslogng** [comment]
By the way, I'm still desperate to get some form of `-Werror` merged ASAP. I've uncovered a few embarrassing bugs/typos yesterday that I will submit later today.

@bazsi @ihrwein  If you don't like any of my proposed workarounds, at least put a comment at each affected include directive that must stay put (the ones altered by your commit and maybe others as well), put a big banner on top of each include header which needs to be included specially, just merge that and forget about it. This is my last input, otherwise :+1:

**Comment by bazsi** [comment]
Python is in a satellite module whose only aim is to bring in Python
support.

This shouldn't plague anything outside modules/python, if it is, we have a
bug.

Iirc we had a similar problem in Zorp where Python plays a much more
central role, and at the end we concluded including Python.h first. It
don't remember problems resulting from this.

We already have a policy to include syslog-ng.h first, but I would never
include python.h from syslog-ng.h, that would pollute our global namespace
a lot.

**Comment by bkil-syslogng** [comment]
Now as I checked, it's not the first occasion when we need to adjust macros:
https://github.com/balabit/syslog-ng/blob/master/modules/afsocket/afinet-dest.c#L39
https://github.com/balabit/syslog-ng/blob/master/modules/afsocket/afsocket-grammar.ym#L59

**Comment by bazsi** [comment]
Yeah, libnet is broken as well.

**Comment by bazsi** [comment]
What do you think of this solution:
https://github.com/balabit/syslog-ng/tree/f/turn-on-Werror-bazsi

**Comment by bkil-syslogng** [comment]
@bazsi If the whole stuff still works correctly, I'm very happy with your proposed patch :+1:

**Comment by bkil-syslogng** [comment]
According to my tests, more warning fixes will be needed, otherwise merging this would break master: https://github.com/bkil-syslogng/syslog-ng/pull/12

**Comment by ihrwein** [comment]
I'm closing this in favor of #871
